include/utl/logging.h

```diff
@@ -57,46 +45,83 @@ inline std::string now() {
   return ss.str();
 }
 
-/// Produce a new log line at the given `level`, with the given prefix `ctx` and
-/// message
-template <typename... Args>
-void log(log_level const level, char const* ctx,
-         fmt::format_string<Args...> fmt_str, Args&&... args) {
-  if (level >= ::utl::s_verbosity) {
-    fmt::print(std::clog, "{time} [{level}] [{ctx}] {msg}\n",
-               fmt::arg("time", now()), fmt::arg("level", to_str(level)),
-               fmt::arg("ctx", ctx),
-               fmt::arg("msg", fmt::format(fmt::runtime(fmt_str),
-                                           std::forward<Args>(args)...)));
+/// Produce a new log line at the given `level`, with the given message
+template <log_level LogLevel, typename... Args>
+struct log {
+  log(const char* ctx, fmt::format_string<Args...> fmt_str,
+      const Args&&... args,
+      std::source_location const& loc = std::source_location::current())
+      : loc_{loc},
+        ctx_{ctx},
+        msg_{fmt::format(fmt_str, std::forward<Args>(args)...)} {}
+
+  ~log() {
+    if (LogLevel >= utl::s_verbosity) {
+#if defined(_WIN32)
+      const char* baseFileName = strrchr(loc_.file_name(), '\\')
+                                     ? strrchr(loc_.file_name(), '\\') + 1
+                                     : loc_.file_name();
+#else
+      // On MacOS, due to a bug with Clang 15, the wrong filename
+      // is retrieved (logging.h instead of the calling file):
+      // https://github.com/llvm/llvm-project/issues/56379
+      const char* baseFileName = strrchr(loc_.file_name(), '/')
+                                     ? strrchr(loc_.file_name(), '/') + 1
+                                     : loc_.file_name();
+#endif
+      fmt::print(std::clog,
+                 "{time} [{level}] [{file}:{line} {fn}] [{ctx}] {msg}\n",
+                 fmt::arg("time", now()), fmt::arg("level", to_str(LogLevel)),
+                 fmt::arg("file", baseFileName), fmt::arg("line", loc_.line()),
+                 fmt::arg("fn", loc_.function_name()), fmt::arg("ctx", ctx_),
+                 fmt::arg("msg", msg_));
+    }
   }
-}
 
-}  // namespace utl
+  void attrs(
+      std::initializer_list<std::pair<std::string_view, std::string_view> >&&
```

Comment:
Maybe add to the "future work" issue that we could support more datatypes by making it a `std::pair<std::string_view, std::variant</* ... all supported value types */>>`. This would require a custom formatter for libfmt but it's doable.

Reply:
Alright, added.